src/_pytest/nodes.py

```diff
@@ -144,8 +144,10 @@ def __init__(
                 self._nodeid += "::" + self.name
 
     @classmethod
-    def from_parent(cls, parent, *, name):
-        return cls._create(parent=parent, name=name)
+    def from_parent(cls, parent, **kw):
```

Comment:
Can you explain this change? The previous signature already didn't allow `config` and `session`, why change to `**kw` here?

Reply:
this is to allow `super().from_parent` - else subclasses would have to use horrendous hacks around the internal '_create' method

Reply:
FWIW, I still think we should not have / need `from_parent` to begin with, but could use raising TypeError and/or type annotations. See https://github.com/pytest-dev/pytest/pull/5975#discussion_r339523282.  I've hoped there to go back to "the simpler pattern" before that PR was merged even.

Reply:
@blueyed from_parent will very likely stay in order to manage the logic parts of node construction

its very likely that `Node.from_parent` will eventually manage passing `config`, `session`, `parent` and `nodeid` to the `Node` constructor

Reply:
IIUC, adding `from_parent` is needed to allow for a *deprecation* of the current way of constructing `Node` and subclasses. 

Again IIUC, the current `Node` constructor (and subclasses) allow for the creation of broken states and configuration, for example `Function` and `Session` objects with different `Config`s, so the idea is for `__init__` to no longer be public API, used only internally.

Perhaps it would help if @blueyed provided a code sample of what the `Node.__init__` method would looke like in his mind?

Reply:
> Perhaps it would help if @blueyed provided a code sample of what the `Node.__init__` method would looke like in his mind?

It could behave as `from_parent` if only `parent` is provided, and ensure that no other (conflicting) kwargs are there (raising a DeprecationWarning (and later TypeError) then, e.g. for when `config` or `session` is passed also).

My main issue here is that it complicates things in general, and requires plugins to adopt it (adding a compat layer there), e.g. pytest-asyncio (https://github.com/pytest-dev/pytest-asyncio/blob/6f72dd9f419f7945906d566a85ba8d50b8d8324d/pytest_asyncio/plugin.py#L39).

> IIUC, adding `from_parent` is needed to allow for a _deprecation_ of the current way of constructing `Node` and subclasses.

Well, the deprecation is for having to use `from_parent` after all.
From what I can tell we could also have deprecations via `__init__` for when an old/unsupported combination/pattern is used.

Reply:
for me a key problem is construction correctness, right now there are just too many things entangled that cant be easily be de-tangled with a constructor without keeping unreasonably smart `__init__` methods -- imho the unreasonable smartness needs to be dissolved or moved into factory functions/methods (like `from_parent`) 

i have been fighting with the details over a long time now as i tried repeatedly to make `FunctionDefinition` an actual part of the collection tree, and it just keeps breaking and breaking in strange places

my goal is to remove those breakage points one by one so i c an put things in place, one of those stumbling blocks are unreasonably smart constructors (they just do way too much)

i'm happy to go for a different solution that works, but so far nobody has show one, as such im slowly clipping away at the things that make future work utterly horrible for me,

as for the constructors as they currently work - im not aware of a way to untangle them without making them inaccessible at once to allow horrendous internal breakages that wont affect users

Reply:
also sorry for the late reply, i missed the comment earlier, @nicoddemus brought it to my attention

Reply:
@RonnyPfannschmidt I can see that it makes sense to get there, but it is bad when plugins have to add compat code for it then, i.e. it would be best to remove it again before releasing it.

As for getting there with type annotations and in small steps see e.g. https://github.com/pytest-dev/pytest/pull/6461.

Reply:
@blueyed its very likely going to stay (so the `__init__` wont have to have logic) and i expect it to take many  months to completely sort things out